plugins/multifundchannel.c

```diff
@@ -0,0 +1,1939 @@
+safely broadcast the tx, so that in case the
+peer disappears, we can recover our funds.
+
+The `fundchannel_complete` command performs the
+negotiation with the peer to sign the initial
+commiteent transactions.
+Only once the `lightningd` has the transactions
+signed does the `fundchannel_complete` command
+return with a success.
+After that point we can `signpsbt`+`sendpsbt`
+the transaction.
+*/
+
+static void
+fundchannel_complete_dest(struct multifundchannel_destination *dest);
+
+static struct command_result *
+perform_fundchannel_complete(struct multifundchannel_command *mfc)
+{
+	unsigned int i;
+
+	plugin_log(mfc->cmd->plugin, LOG_DBG,
+		   "mfc %"PRIu64": parallel fundchannel_complete.",
+		   mfc->id);
+
+	mfc->pending = tal_count(mfc->destinations);
+
+	for (i = 0; i < tal_count(mfc->destinations); ++i)
+		fundchannel_complete_dest(&mfc->destinations[i]);
+
+	assert(mfc->pending != 0);
+	return command_still_pending(mfc->cmd);
+}
+
+static struct command_result *
+fundchannel_complete_ok(struct command *cmd,
+			const char *buf,
+			const jsmntok_t *result,
+			struct multifundchannel_destination *dest);
+static struct command_result *
+fundchannel_complete_err(struct command *cmd,
+			 const char *buf,
+			 const jsmntok_t *error,
+			 struct multifundchannel_destination *dest);
+
+static void
+fundchannel_complete_dest(struct multifundchannel_destination *dest)
+{
+	struct multifundchannel_command *mfc = dest->mfc;
+	struct command *cmd = mfc->cmd;
+	struct out_req *req;
+
+	plugin_log(mfc->cmd->plugin, LOG_DBG,
+		   "mfc %"PRIu64", dest %u: fundchannel_complete %s.",
+		   mfc->id, dest->index,
+		   node_id_to_hexstr(tmpctx, &dest->id));
+
+	req = jsonrpc_request_start(cmd->plugin,
+				    cmd,
+				    "fundchannel_complete",
+				    &fundchannel_complete_ok,
+				    &fundchannel_complete_err,
+				    dest);
+	json_add_node_id(req->js, "id", &dest->id);
+	json_add_string(req->js, "txid",
+			type_to_string(tmpctx, struct bitcoin_txid,
+				       mfc->txid));
+	json_add_num(req->js, "txout", dest->outnum);
+
+	send_outreq(cmd->plugin, req);
+}
+
+static struct command_result *
+fundchannel_complete_done(struct multifundchannel_destination *dest);
+
+static struct command_result *
+fundchannel_complete_ok(struct command *cmd,
+			const char *buf,
+			const jsmntok_t *result,
+			struct multifundchannel_destination *dest)
+{
+	struct multifundchannel_command *mfc = dest->mfc;
+	const jsmntok_t *channel_id_tok;
+
+	plugin_log(mfc->cmd->plugin, LOG_DBG,
+		   "mfc %"PRIu64", dest %u: fundchannel_complete %s done.",
+		   mfc->id, dest->index,
+		   node_id_to_hexstr(tmpctx, &dest->id));
+
+	channel_id_tok = json_get_member(buf, result, "channel_id");
+	if (!channel_id_tok)
+		plugin_err(cmd->plugin,
+			   "fundchannel_complete no channel_id: %.*s",
+			   json_tok_full_len(result),
+			   json_tok_full(buf, result));
+	dest->channel_id = json_strdup(mfc, buf, channel_id_tok);
```

Comment:
since you don't change the state here, you wouldn't catch if (for some reason) a command failed to complete when you check for 'finishedness' in `after_fundchannel_complete`

Reply:
I do not understand.